File: External/SpectraCGate_x86/SDK/samples/net/basic/send_mt.cs
/*
		Copyright (c) 2014, MOEX. All rights reserved.
		
		Plaza-2 Client Gate API Usage Sample.
		Replication DataStream Client.
		
		All the software and documentation included in this and any
		other MOEX CGate Releasese is copyrighted by MOEX.
		
		Redistribution and use in source and binary forms, with or without
		modification, are permitted only by the terms of a valid
		software license agreement with MOEX.
		
		THIS SOFTWARE IS PROVIDED "AS IS" AND MICEX-RTS DISCLAIMS ALL WARRANTIES
		EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
		NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR
		PURPOSE.  MICEX-RTS DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE
		UNINTERRUPTED OR ERROR-FREE.  MICEX-RTS SHALL NOT, UNDER ANY CIRCUMSTANCES, BE
		LIABLE TO LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR
		INDIRECT DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE
		TRANSACTIONS CONTEMPLATED HEREUNDER, EVEN IF MICEX-RTS HAS BEEN APPRISED OF
		THE LIKELIHOOD OF SUCH DAMAGES.
*/

/* 
	Send_mt - многопоточный пример посылки заявки. (Примечание: компилируется только под компиляторами, поддерживающими C+
	+11.). В треде 1 посылаются заявки.В треде 2 обрабатываются reply на посылаемые заявки.

	Send_mt - a multistream example of sending order. (Attention! This example can be compiled only by compilers supporting C++11.)
	Stream 1 contains orders. Stream 2 contains replies for the orders sent.
*/

using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;

using System.Runtime.InteropServices;

using ru.micexrts.cgate;
using ru.micexrts.cgate.message;

namespace send
{
	
	class MessageBase
	{
		protected static bool bExit = false;
		static private Object thisLock = new Object();

		static public bool BExit
		{
			get
			{
				lock (thisLock)
				{
					return BExit;
				}
			}
			set 
			{
				lock (thisLock)
				{
					bExit = value;
				}
			}
		}
	}
	
	class MessageReplyListener : MessageBase
	{
		const int defaultTimeout = 1;
		int cnt = 0;
		public MessageReplyListener(string connSettings, string lsnSettings)
		{
			conn = new Connection(connSettings);
			listener = new Listener(conn, lsnSettings);
			listener.Handler += new Listener.MessageHandler(this.DataEvent);
		}
	
		public void Close()
		{
			listener.Close();
			listener.Dispose();
			conn.Close();
			conn.Dispose();
		}
	
	
		void processListener()
		{
			State state = listener.State;
			if (state == State.Error)
				listener.Close();
			else if (state == State.Closed)
				listener.Open();
		}
	
		void processConnection()
		{
			State state = conn.State;
			if (state == State.Error)
				conn.Close();
			else if (state == State.Closed)
				conn.Open();
			else if (state == State.Active)
			{
				conn.Process(defaultTimeout);
				processListener();
			}
		}
	
		public void run()
		{
			DateTime lastTime = DateTime.Now;
			TimeSpan oneSecond = new TimeSpan(0, 0, 1);
			while(!bExit)
			{
				try
				{
					processConnection();
				}
				catch(CGateException e)
				{
					CGate.LogError(String.Format("Listener run: cgate exception %{0}",e.ToString()));
				}

				TimeSpan span = DateTime.Now - lastTime;
				if (span >= oneSecond)
				{
					CGate.LogInfo("Receive: messages per second " + cnt);
					cnt = 0;
					lastTime = DateTime.Now;
				}
			}
			CGate.LogInfo("Listener: finished");
		}
	
		public int DataEvent(Connection conn, Listener listener, Message msg)
		{
			if (msg.Type == MessageType.MsgData)
			{
				CGate.LogInfo(String.Format("message reply dump: {0}", msg));
				cnt++;
			}
			return 0;
		}
	
		public Connection getConnection()
		{
			return conn;
		}
	
		Connection conn;
		Listener listener;
	}
	
	class MessagePublisher : MessageBase
	{
	
		int cnt = 0;
		public MessagePublisher(Connection conn, string pubSettings)
		{
			this.conn = conn;
			publisher = new Publisher(conn, pubSettings);
		}
	
		public void Close()
		{
			publisher.Close();
			publisher.Dispose();
		}
	
		private static void sendRequest(Publisher pub, int counter)
		{
			SchemeDesc schemeDesc = pub.Scheme;
			MessageDesc messageDesc = schemeDesc.Messages[0]; //AddMM
			FieldDesc fieldDesc = messageDesc.Fields[3];
	
			Message sendMessage = pub.NewMessage(MessageKeyType.KeyName, "FutAddOrder");
	
			DataMessage smsg = (DataMessage)sendMessage;
			smsg.UserId = (uint)counter;
			smsg["broker_code"].set("HB00");
			smsg["isin"].set("RTS-6.12");
			smsg["client_code"].set("000");
			smsg["type"].set(1);
			smsg["dir"].set(1);
			smsg["amount"].set(1);
			smsg["price"].set("1700");
			smsg["ext_id"].set(0);
	
			CGate.LogInfo(String.Format("posting message dump: {0}", sendMessage));
			pub.Post(sendMessage, PublishFlag.NeedReply);
			sendMessage.Dispose();
		}
	
		void DoLogic()
		{
			try
			{
				sendRequest(publisher, cnt);
				cnt = cnt + 1;
			}
			catch (CGateException e)
			{
				CGate.LogError(String.Format("Publisher do logic: cgate exception %{0}", e.ToString()));
				cnt = cnt - 1;
			}
		}
	
		void processPublisher()
		{
			State state = publisher.State;
			if (state == State.Error)
				publisher.Close();
			else if (state == State.Closed)
				publisher.Open();
			else if (state == State.Active)
				DoLogic();
		}
	
		public void run(Object arg)
		{
			CountdownEvent startEvent = (CountdownEvent)arg;
			startEvent.Signal();
			startEvent.Wait();

			DateTime lastTime = DateTime.Now;
			TimeSpan oneSecond = new TimeSpan(0, 0, 1);
			CGate.LogInfo("Thread started");
			while (!bExit)
			{
				try
				{
					processPublisher();
				}
				catch (CGateException e)
				{
					CGate.LogError(String.Format("Publisher run: cgate exception %{0}", e.ToString()));
				}

				TimeSpan span = DateTime.Now - lastTime;
				if (span >= oneSecond)
				{
					CGate.LogInfo("Post: messages per second " + cnt);
					cnt = 0;
					lastTime = DateTime.Now;
				}
			}
			CGate.LogInfo("Publisher: finished");
		}
		
		Connection conn;
		Publisher publisher;
	}
		
	class SendMt
	{
	
		static MessageReplyListener replyListener = null;
		const int nThreads = 2;
		static CountdownEvent startEvent = new CountdownEvent(nThreads);
	
		public static void ConsoleCancelEventHandler(object sender, ConsoleCancelEventArgs e)
		{
			MessageBase.BExit = true;
			CGate.LogInfo(" - - - BREAK - - - ");
		}
	
		public static void Main(string[] args)
		{
			Console.CancelKeyPress += ConsoleCancelEventHandler;
			CGate.Open("ini=netsend_mt.ini;key=11111111");
			CGate.LogInfo("test .Net log.");

			MessageReplyListener replyListener = new MessageReplyListener("p2tcp://127.0.0.1:4001;app_name=send_mt", "p2mqreply://;ref=srvlink0");
			MessagePublisher[] publishers = new MessagePublisher[nThreads];
			for (int i = 0; i < nThreads; ++i)
			{
				String name = String.Format("p2mq://FORTS_SRV;category=FORTS_MSG;name=srvlink{0};timeout=5000;scheme=|FILE|forts_messages.ini|message", i);
				publishers[i] = new MessagePublisher(
					replyListener.getConnection(), 
					String.Format("p2mq://FORTS_SRV;category=FORTS_MSG;name=srvlink{0};timeout=5000;scheme=|FILE|forts_messages.ini|message", i));
			}
			
			try
			{
				Thread replyThread = new Thread(new ThreadStart(replyListener.run));
				replyThread.Start();

				Thread[] sendThreads = new Thread[nThreads];
				for (int i = 0; i < nThreads; ++i)
				{
					sendThreads[i] = new Thread(new ParameterizedThreadStart(publishers[i].run));
					sendThreads[i].Start(startEvent);
				}

				replyThread.Join();
				for (int i = 0; i < nThreads; ++i)
				{
					sendThreads[i].Join();
				}
				CGate.LogInfo("All threads joined");
			}
			catch (CGateException e)
			{
				CGate.LogError(String.Format("Main: cgate exception %{0}", e.ToString()));
			}

			for (int i = 0; i < nThreads; ++i)
			{
				publishers[i].Close();
			}
			replyListener.Close();
			
			CGate.LogInfo("Finish");
			CGate.Close();
		}
	}
}

File: External/SpectraCGate_x64/SDK/samples/c/concurrent/send_mt.cpp
/*
        Copyright (c) 2012, MICEX-RTS. All rights reserved.

        Plaza-2 Client Gate API Usage Sample.
        Messages sender.

        All the software and documentation included in this and any
        other MICEX-RTS CGate Releasese is copyrighted by MICEX-RTS.

        Redistribution and use in source and binary forms, with or without
        modification, are permitted only by the terms of a valid
        software license agreement with MICEX-RTS.

        THIS SOFTWARE IS PROVIDED "AS IS" AND MICEX-RTS DISCLAIMS ALL WARRANTIES
        EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION, ANY IMPLIED WARRANTIES OF
        NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A PARTICULAR
        PURPOSE.  MICEX-RTS DOES NOT WARRANT THAT USE OF THE SOFTWARE WILL BE
        UNINTERRUPTED OR ERROR-FREE.  MICEX-RTS SHALL NOT, UNDER ANY CIRCUMSTANCES, BE
        LIABLE TO LICENSEE FOR LOST PROFITS, CONSEQUENTIAL, INCIDENTAL, SPECIAL OR
        INDIRECT DAMAGES ARISING OUT OF OR RELATED TO THIS AGREEMENT OR THE
        TRANSACTIONS CONTEMPLATED HEREUNDER, EVEN IF MICEX-RTS HAS BEEN APPRISED OF
        THE LIKELIHOOD OF SUCH DAMAGES.
*/

/* 
	Send_mt - многопоточный пример посылки заявки. (Примечание: компилируется только под компиляторами, поддерживающими C+
	+11.). В треде 1 посылаются заявки.В треде 2 обрабатываются reply на посылаемые заявки.

	Send_mt - a multistream example of sending order. (Attention! This example can be compiled only by compilers supporting C++11.)
	Stream 1 contains orders. Stream 2 contains replies for the orders sent.
*/

#include <thread>
#include <memory>
#include <vector>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <condition_variable>

#ifdef _WIN32
#include <windows.h>
#endif//_WIN32

//#include <assert.h>
#define CG_SHORT_NAMES

#include <cgate.h>
#include "cgate.hpp"
using namespace cgate; 

#include "forts_messages.h"
#include "send_mt_utils.h"

// helper variable used as a notification flag to exit
static volatile int bExit = 0;
const int numPubs = 5;
static volatile int id = 0;

#ifdef _WIN32
BOOL WINAPI InterruptHandler(DWORD reason)
{
	printf("\n----BREAK----\n");
	bExit = 1;
	return 1;
}
#else
#include <unistd.h>
void Sleep(int msec)
{
	usleep(msec * 1000);
}
#endif//_WIN32

const uint32_t defaultTimeOut = 1;

class MessageReplyListener : public Listener::IEventsListener
{
public:
	MessageReplyListener(const std::string& connSettings, const std::string& lsnSettings) :
	  _connSettings(connSettings),
	  _lsnSettings(lsnSettings),
	  _cnt(0),
	  _tmr(0)
	{
		_conn.reset(new Connection(connSettings));
		_lsn.reset(new Listener(*_conn, lsnSettings, this));
	}

	inline void processListener()
	{
		State state = _lsn->getState();
		if (state == State::Error)
			_lsn->close();
		if (state == State::Closed)
			_lsn->open(_lsnSettings);
	}

	inline void processConnection()
	{
		State state = _conn->getState();
		if (state == State::Error)
			_conn->close();
		if (state == State::Closed)
			_conn->open(_connSettings);
		if (state == State::Active)
		{
			_conn->process(defaultTimeOut);
			processListener();
		}
	}

	void run()
	{
		_tmr = cg_usec_clock();
		while(!bExit)
		{
			try
			{
				processConnection();
			}
			catch(Exception& e)
			{
				if (e.code() != CG_ERR_TIMEOUT)
					CGate::logError("Listener run: cgate exception %d, %s", e.code(), e.what().c_str());
			}
		}
		CGate::logInfo("Listener: finished");
	}

	uint32_t event(cg_conn_t* conn, cg_listener_t* listener, struct cg_msg_t* msg)
	{
		if (msg->type == CG_MSG_DATA)
		{
			cg_msg_data_t* msgData = reinterpret_cast<cg_msg_data_t*>(msg);
			if (cg_usec_clock() - _tmr >= 1000000)
			{
				CGate::logInfo("Receive: Msg per sec %d", _cnt);
				_tmr = cg_usec_clock();
				_cnt = 0;
			}
			++_cnt;
			//CGate::logInfo("Msg received reply for %d", msgData->user_id);
		}
		return CG_ERR_OK;
	}

	const std::shared_ptr<Connection>& getConnection()
	{
		return _conn;
	}

private:
	MessageReplyListener& operator=(const MessageReplyListener&);
	MessageReplyListener(const MessageReplyListener&);

private:
	std::shared_ptr<Connection> _conn;
	std::shared_ptr<Listener> _lsn;
	std::string _connSettings;
	std::string _lsnSettings;
	uint32_t _cnt;
	uint64_t _tmr;

};

class MessagePublisher
{
public:
	MessagePublisher() :
	    _cnt(0),
	    _tmr(0),
	    _userId(0)
	{
	}
	
	MessagePublisher(const std::shared_ptr< Connection>& conn, const std::string& pubSettings) :
	  _pubSettings(pubSettings),
	  _cnt(0),
	  _tmr(0),
	  _userId(0)
	{
		_conn = conn;
		_pub.reset(new Publisher(*_conn.get(), pubSettings));
	}

	void doLogic()
	{
		try
		{
			struct FutAddOrder* ord ;
			auto msg = _pub->newMessageByName("FutAddOrder");
			MessageData* msgData = dynamic_cast<MessageData*>(msg.get());
			if (msgData == nullptr)
				throw Exception(CG_ERR_INTERNAL);
			msgData->setUserId(++_userId);
			ord = (struct FutAddOrder*)msg->getData();
			strcpy(ord->broker_code, "HB00");
			strcpy(ord->isin, "RTS-6.12");
			strcpy(ord->client_code, "000");
			ord->type = 1;
			ord->dir = 1;
			ord->amount = 1;
			strcpy(ord->price, "1700");
			ord->ext_id = 0;
			uint64_t t = cg_usec_clock();
			if ( uint64_t(t - _tmr) >= 1000000LL)
			{
				CGate::logInfo("Send: Msg per sec %d", _cnt);
				_tmr = t;
				_cnt = 0;
			}
			++_cnt;
			_pub->post(msg.get(), CG_PUB_NEEDREPLY);
			//CGate::logInfo("Msg after send %d", _cnt);
		}
		catch(Exception& e)
		{
			CGate::logError("Publisher do logic: cgate exception %d, %s", e.code(), e.what().c_str());
			--_cnt;
		}
	}

	void processPublisher()
	{
		State state = _pub->getState();
		if (state == State::Error)
			_pub->close();
		if (state == State::Closed)
			_pub->open(_pubSettings);
		if (state == State::Active)
			doLogic();
	}

	void run()
	{
		_tmr = cg_usec_clock();
		while(!bExit)
		{
			try
			{
				State state = _conn->getState();
				if (state == State::Active)
				{
					processPublisher();
				}
			}
			catch(Exception& e)
			{
				CGate::logError("Publisher run: cgate exception %d, %s", e.code(), e.what().c_str());
			}
		}
		CGate::logInfo("Publisher: finished");
	}
	
	static bool pred()
	{
		return id == numPubs;
	}

	static void createAndRun(const std::shared_ptr<Connection>& conn, int i, std::unique_ptr<MessagePublisher>* publisher, std::mutex* cv_m, std::condition_variable* cv)
	{
		char settings[256];
		sprintf(settings, "p2mq://FORTS_SRV;category=FORTS_MSG;name=srvlink%d;timeout=5000;scheme=|FILE|forts_messages.ini|message", i);
		{
			{
				std::unique_lock<std::mutex> lk(*cv_m);
				id++;
				(*cv).wait(lk, pred);
				(*cv).notify_all();
			}
		}
		CGate::logInfo("Thread %d is ready to start\n", i);
		publisher->reset(new MessagePublisher(conn, settings));
		publisher->get()->run();
	}

private:
	MessagePublisher& operator=(const MessagePublisher&);
	MessagePublisher(const MessagePublisher&);
private:
	std::shared_ptr<Connection> _conn;
	std::shared_ptr<Publisher> _pub;
	std::string _pubSettings;
	uint32_t _userId;
	uint32_t _cnt;
	uint64_t _tmr;
};


int main()
{	
#ifdef _WIN32
	SetConsoleCtrlHandler(InterruptHandler, 1);
#endif//_WIN32
	try
	{
		CGate env("ini=send_mt.ini;key=11111111;");
		std::unique_ptr<MessageReplyListener> replyListener;
		replyListener.reset(new MessageReplyListener("p2tcp://127.0.0.1:4001;app_name=test_send", "p2mqreply://;ref=srvlink0"));
		std::thread threadListener(&MessageReplyListener::run, replyListener.get());
		Sleep(100);

		std::condition_variable cv;
		std::mutex cv_m;


		std::vector< std::unique_ptr<MessagePublisher> > publishers;
		for(int i = 0; i < numPubs; ++i)
		{
		    publishers.push_back(std::unique_ptr<MessagePublisher>(new MessagePublisher()));
		}

		std::vector<std::thread> threads;
		threads.reserve(numPubs);
		int id = 0;

		for(int i = 0; i < numPubs; ++i)
		{
			threads.push_back(std::thread(MessagePublisher::createAndRun, replyListener->getConnection(), i, &publishers[i], &cv_m, &cv));
		}

		threadListener.join();
		for(int i = 0; i < numPubs; ++i)
		{
			threads[i].join();
		}
		CGate::logInfo("All threads joined");
	}
	catch(Exception& e)
	{
		CGate::logError("Main: cgate exception %d, %s", e.code(), e.what().c_str());
	}
	return 0;
}
